### DISCUSSION TITLE
How to create a code space for final project

### DISCUSSION THREAD

**Discussion by oye-bobs** [cu-esiil-edu/2024-data-short-course#97] [Q&A]
Hi @eculler @nquarder , For My final project, I would like to create a notebook to teach students at my institution (University of Benin) what reproducible science is, how to collaborate on github, and how to access and explore climate data for different states in Nigeria, How do I create a code space to this effect?

**Comment by nquarder** [comment]
This is a great idea @oye-bobs and think it will be wonderful for you to teach these topics to your colleagues and classmates at University of Benin. 

I would suggest that you create a new repository for this project. You can create a copy of the `.devcontainer` that you used for the vegetation notebook, as well as create a `README.md`, `index.md`, License, and a `ipynb` Jupyter Notebook (that you can convert to html) for your work. 

You might also consider creating a folder for images (img) you want add to your notebook or page. I would consider creating a GitHub page for this repository so you can easily share it with others who may not be regular users of GitHub but that you want to show your work off to. 

Once you have your repository created along with the associated files that you want to include, then you can create a codespace for this project and begin adding the content you want to the Jupyter Notebook. 

A couple things to consider here, too. You have unlimited codespace minutes if you use our cu-esiil-edu organization to host the repository. As an ESIIL community member, you are welcome to do this!

If you are using your oye-bobs account to host the repository, which I might recommend, then you will be limited in the number of codespace minutes you have each month. You can upgrade to pro account ($4 USD/month, I think?) and you can increase the minutes you have for codespaces. 

One difference will be in the URL attached the repository. If you host it, the URL will look something like: 

`https://github.com/oye-bobs/<project-repo-name-here>`

If you host it on the cu-esiil-edu org, then the URL will look more like:

`https://github.com/cu-esiil-edu/<project-repo-name-here>`

One other solution would be to create the repo in the cu-esiil-edu org, use our codespace minutes to do all of your work, and then FORK the repository over to your personal account. That way you won't use up your codespace minutes for this project. 

Finally, @oye-bobs, if you are interested in doing more advanced work that will require lots of computational minutes each month, we are able to help set you up with an account in [CyVerse](https://cyverse.org/tools). This would give you access to more compute, more storage, and more open science tools. Think about this and if you want to chat more we can. An an ESIIL community member you have free access to these tools. For the project, I would say work in GitHub/codespaces, and we can move to CyVerse later if that is something you want to explore.

**Reply by oye-bobs** [reply]
Hi @nquarder , I just created the code space after uploading both the dev container and the gitignore files,  but selecting a kernel/intepreter is difficult, I haven't been able to do that, How do I fix that problem? Do I have to install a python intepreter within the code space?

**Comment by nquarder** [comment]
Hi @oye-bobs: 

A couple of things. 

1.) Make sure you call your file `.devcontainer.json`

2.) You can copy of the contents from the container we used for the vegetation assignment. See below:

```
{
    "name": "Earth Analytics Python Container",
    "image": "registry.hub.docker.com/earthlab/earth-analytics-python-env:latest",
    "extensions": [
        "ms-python.python",
        "ms-toolsai.jupyter",
        // Use black to format code (right click)
        "ms-python.black-formatter",
    ],
    "settings": {
        // Display line length guides
        "editor.rulers": [72, 79],
        // Turn off annoying overzealous autocomplete
        "editor.acceptSuggestionOnEnter": "off",
        // Formatting tools
        "editor.defaultFormatter": "ms-python.black-formatter",
        "python.formatting.blackArgs": [
            "--line-length=79",
            "--experimental-string-processing"
        ],        
        // Set the default python to the conda install
        "python.defaultInterpreterPath": "/opt/conda/bin/python"
    }
}
```

Are you able to post a screenshot or the code from your container file? Also when you launch the notebook and Select the Kernel you may be asked to install Python extensions. I would go ahead and do that if prompted. Try this and see if this helps! -NQ

**Reply by oye-bobs** [reply]
Hello @nquarder , here is what my .json container looks like, It's the one from the vegetation notebook. ![12](https://github.com/cu-esiil-edu/2024-data-short-course/assets/33554816/2e811e09-60a4-4221-8c49-25a3e50db93f)

**Reply by nquarder** [reply]
Okay @oye-bobs you need to rename this file to `.devcontainer.json` . Notice the extra `.` to the left of the `d` in `devcontainer`? You need to add that to your file name and this should help fix this issue. -NQ

**Comment by oye-bobs** [comment]
@nquarder it also seems like the codespace allocation has reached its limit.

**Reply by nquarder** [reply]
Hi @oye-bobs - can you go to github.com/codespace and scroll down to the bottom of the screen. Take a screen shot and post here if you can. I imagine you have old codespaces that are using up your monthly storage allotment. Typically, you can safely delete a codespace once you have committed your changes and are no longer regularly working on the repository.

You can also apply for a free version of GitHub pro if you are a student or educator. Read more about that here: 

https://docs.github.com/en/billing/managing-the-plan-for-your-github-account/discounted-plans-for-github-accounts

It will give you more compute minutes and storage each month. Even with a Pro account, it is still good to get in the habit of deleting a codespace when you are done working for the day, as long as you remember to commit your changes, you work will still be there. 

I'm cc'ing @Sordhiq here too as they had a similar question about Codespaces.

**Reply by oye-bobs** [reply]
Hi @nquarder , Here is the screen shot you asked for. 
![13](https://github.com/cu-esiil-edu/2024-data-short-course/assets/33554816/e78c6c1c-0c3a-450d-a7e3-1159cac218c2)


However, there is only one other codespace active because I have deleted the others.

**Reply by nquarder** [reply]
Hi @oye-bobs - I was running into this same issue late last week, though it now appears that my account usage has been reset effective this morning, May 20th 2024? Has your's been reset? If not then it may be June 01 when the reset will take place, at which point you'll be able to use Codespaces again with your free, personal account. 

Also, I would say it is okay to delete the 01-get-started-oye-bobs codespace (fuzzy waddle) once you are done working on that piece. You can create a new Codespace and you work will still be there as long as you are committing your changes. This will save you from using up your monthly storage allotment. 

One other thing you may want to consider is looking into creating an education organization account for the UniBenOpenSciEd program. This would give you access to additional resources within GitHub Education as a student/educator/researcher. 

From this Quickstart menu you can read about the free resources available to students and educators:

https://docs.github.com/en/education/quickstart

It might be fun to create this organization as part of your project. You would be the maintainer for now and then invite others to contribute as you introduce them to open/reproducible science. 

If not then it's okay to wait a few days until the end of the month to pick up with the short course.

**Reply by oye-bobs** [reply]
Hi @nquarder I just create a new github organisation here is the link (https://github.com/UnibenEarthDataScience) How do I begin creating the project in the new organisation?

**Reply by oye-bobs** [reply]
@nquarder even after creating a new organisation, I was still unable to use the codespaces. it gives this error:
![Screenshot 2024-05-27 165139](https://github.com/cu-esiil-edu/2024-data-short-course/assets/33554816/78c90e30-179a-4777-a283-bcb8454ade92)

**Reply by nquarder** [reply]
okay @oye-bobs so sorry for the delay in my response here. i wonder now that it is the beginning of a new month, are you now able to get back going in a codespace? if so you should be able to proceed just as you did with your personal account. you can create a repository for your project and then the workflow will be similar. let me know if you want to set up a meeting to chat. i will also be announcing short course office hours beginning again in a couple of weeks. stay tuned! 👍

**Reply by oye-bobs** [reply]
Hi @nquarder, no worries! I understand how busy you've been; I've been busy too. I just checked, and the codespace is working properly now.

**Reply by oye-bobs** [reply]
@nquarder, I have finished creating the codespace. Here is the link: [Link](https://github.com/cu-esiil-edu/UniBenOpenSciEd). Could you please review it?

Additionally, I wanted to inquire about the ESIIL badge for successfully completing the short course.

**Reply by nquarder** [reply]
This is great work @oye-bobs and thank you for sharing this work with the community members at UniBEN. We are still working on getting the badges complete but will have them ready to share next week sometime. I will reach out to you when they are ready. We will post to our ESIIL LinkedIn and announce you have completed the short course and will tag you if you are comfortable with that?

I'm also cc'ing @DharaniSureshBabu and @gauravchakrabarty who are also done with the requirements and will be receiving the badge/certificate of completion next week sometime. 

Thanks again everyone for your hard work! We are very proud of you and your accomplishments in this course :smile:

**Reply by oye-bobs** [reply]
Thank you, @nquarder and @eculler, for sharing your earth data science knowledge! This course was a game-changer. As a recent Geology and Earth Science graduate, it opened my eyes to the boundless possibilities in the fields of Geology, Earth Sciences, Data Science, and Earth Data Science. It also equipped me with the skills to take my career to the next level. I'm excited to see what the future holds!

Thank you once again, and yes, you can tag me in your LinkedIn post.

**Reply by gauravchakrabarty** [reply]
> This is great work @oye-bobs and thank you for sharing this work with the community members at UniBEN. We are still working on getting the badges complete but will have them ready to share next week sometime. I will reach out to you when they are ready. We will post to our ESIIL LinkedIn and announce you have completed the short course and will tag you if you are comfortable with that?
> 
> I'm also cc'ing @DharaniSureshBabu and @gauravchakrabarty who are also done with the requirements and will be receiving the badge/certificate of completion next week sometime.
> 
> Thanks again everyone for your hard work! We are very proud of you and your accomplishments in this course 😄

Waiting eagerly for the certificate !!
@nquarder @eculler
